hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/JetMetricsService.java

```diff
@@ -259,4 +267,40 @@ private void logError(String name, Object value, MetricsPublisher publisher, Exc
             logger.fine("Error publishing metric to: " + publisher.name() + ", metric=" + name + ", value=" + value, e);
         }
     }
+
+    /**
+     * Internal publisher which notifies the {@link JobExecutionService} about
+     * latest metric values.
+     */
+    public static class InternalJobMetricsPublisher implements MetricsPublisher {
+
+        private final JobExecutionService jobExecutionService;
+
+        private final Map<String, Long> metrics = new HashMap<>();
+
+        public InternalJobMetricsPublisher(JobExecutionService jobExecutionService) {
+            this.jobExecutionService = jobExecutionService;
+        }
+
+        @Override
+        public void publishLong(String name, long value) {
+            metrics.put(name, value);
```

Comment:
We can directly collect to `Map<ExecId, Map<MetricName, Long>>` and in `whenComplete` just use that maps.

Reply:
Yes, done, see previous.